File: src/storage/markdown_store.py
"""
Markdown file store on disk.

Layout (v2): ``diary_root/YYYY/MM/<id>.md``
Day context: ``diary_root/YYYY/MM/<date>.day.json``
"""
from __future__ import annotations

import logging
import re
import shutil
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from storage.day_store import DayContext, DayStore
from utils.paths import diary_md_relpath, diary_md_relpath_for_id

logger = logging.getLogger("diary.storage.markdown")

FRONT_MATTER_RE = re.compile(
    r"^---\s*\n(.*?)\n---\s*\n(.*)$",
    re.DOTALL,
)
DATE_FILE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}\.md$")
UUID_RE = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)


@dataclass
class EntryFrontMatter:
    date: str = ""
    title: str = ""
    id: str = ""
    created_at: str = ""
    updated_at: str = ""
    writing_duration_sec: int = 0
    tags: list[str] | None = None
    pinned: bool = False
    # Legacy v1 fields (read during migration only)
    location: str = ""
    weather: str = ""
    temp_c: Optional[float] = None
    context_source: str = ""
    context_updated_at: str = ""


class MarkdownStore:
    def __init__(self, diary_root: Path, assets_root: Path | None = None) -> None:
        self.diary_root = diary_root
        self.assets_root = assets_root
        self.diary_root.mkdir(parents=True, exist_ok=True)

    def path_for(self, entry_id: str, entry_date: str) -> Path:
        return self.diary_root / diary_md_relpath_for_id(entry_date, entry_id)

    def read(self, entry_id: str, entry_date: str) -> str:
        path = self.path_for(entry_id, entry_date)
        if not path.exists():
            return ""
        text = path.read_text(encoding="utf-8")
        body, _ = self._parse(text)
        return body

    def read_raw(self, entry_id: str, entry_date: str) -> str:
        path = self.path_for(entry_id, entry_date)
        if not path.exists():
            return ""
        return path.read_text(encoding="utf-8")

    def read_front_matter(self, entry_id: str, entry_date: str) -> EntryFrontMatter:
        path = self.path_for(entry_id, entry_date)
        if not path.exists():
            return EntryFrontMatter(date=entry_date, id=entry_id)
        text = path.read_text(encoding="utf-8")
        _, fm = self._parse(text)
        if not fm.date:
            fm.date = entry_date
        if not fm.id:
            fm.id = entry_id
        return fm

    def write(
        self,
        entry_id: str,
        entry_date: str,
        body: str,
        title: str = "",
        *,
        created_at: str = "",
        updated_at: str = "",
        writing_duration_sec: int = 0,
        tags: list[str] | None = None,
        pinned: bool = False,
    ) -> str:
        """Write markdown file. Returns relative path from diary_root."""
        rel = diary_md_relpath_for_id(entry_date, entry_id)
        path = self.diary_root / rel
        path.parent.mkdir(parents=True, exist_ok=True)
        title = title or entry_date
        lines = [
            "---",
            f"date: {entry_date}",
            f"title: {self._yaml_escape(title)}",
            f"id: {entry_id}",
        ]
        if created_at:
            lines.append(f"created_at: {created_at}")
        if updated_at:
            lines.append(f"updated_at: {updated_at}")
        if writing_duration_sec:
            lines.append(f"writing_duration_sec: {writing_duration_sec}")
        if tags:
            joined = ", ".join(self._yaml_escape(t) for t in tags)
            lines.append(f"tags: [{joined}]")
        if pinned:
            lines.append("pinned: true")
        lines.append("---")
        content = "\n".join(lines) + "\n\n" + body.rstrip() + "\n"
        path.write_text(content, encoding="utf-8")
        logger.debug("Wrote markdown %s", path)
        return rel

    def write_raw(self, entry_id: str, entry_date: str, markdown: str) -> str:
        rel = diary_md_relpath_for_id(entry_date, entry_id)
        path = self.diary_root / rel
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(markdown, encoding="utf-8")
        return rel

    def exists(self, entry_id: str, entry_date: str) -> bool:
        return self.path_for(entry_id, entry_date).exists()

    def list_note_ids(self) -> list[tuple[str, str]]:
        """Scan ``*.md`` files whose stem is a UUID. Returns (id, date) pairs."""
        found: list[tuple[str, str]] = []
        if not self.diary_root.exists():
            return found
        for path in sorted(self.diary_root.rglob("*.md")):
            if DATE_FILE_RE.match(path.name):
                continue
            stem = path.stem
            if not UUID_RE.match(stem):
                continue
            text = path.read_text(encoding="utf-8")
            _, fm = self._parse(text)
            entry_date = fm.date or ""
            if not entry_date:
                continue
            found.append((stem, entry_date))
        return found

    def migrate_v1_layout(self, day_store: DayStore) -> int:
        """
        Convert ``YYYY-MM-DD.md`` files and date-keyed assets to id-keyed layout.
        Returns count of migrated notes.
        """
        if not self.diary_root.exists():
            return 0
        count = 0
        for path in sorted(self.diary_root.rglob("*.md")):
            if not DATE_FILE_RE.match(path.name):
                continue
            entry_date = path.stem
            text = path.read_text(encoding="utf-8")
            body, fm = self._parse(text)
            entry_id = fm.id or str(uuid.uuid4())
            fm.id = entry_id
            fm.date = fm.date or entry_date

            day_ctx = self._extract_legacy_context(text, entry_date)
            if day_ctx and (
                day_ctx.location or day_ctx.weather or day_ctx.temp_c is not None
            ):
                day_store.merge_write(day_ctx)

            new_body = body.replace(f"assets/{entry_date}/", f"assets/{entry_id}/")
            dest = self.path_for(entry_id, fm.date)
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_text(
                self._render_v2(new_body, fm),
                encoding="utf-8",
            )

            if self.assets_root is not None:
                old_assets = self.assets_root / entry_date
                if old_assets.is_dir():
                    new_assets = self.assets_root / entry_id
                    new_assets.mkdir(parents=True, exist_ok=True)
                    for ap in old_assets.iterdir():
                        if ap.is_file():
                            target = new_assets / ap.name
                            if not target.exists():
                                shutil.move(str(ap), str(target))
                    try:
                        old_assets.rmdir()
                    except OSError:
                        pass

            path.unlink(missing_ok=True)
            count += 1
            logger.info("Migrated v1 note %s → %s", entry_date, entry_id)
        return count

    def _render_v2(self, body: str, fm: EntryFrontMatter) -> str:
        title = fm.title or fm.date or "untitled"
        lines = [
            "---",
            f"date: {fm.date}",
            f"title: {self._yaml_escape(title)}",
            f"id: {fm.id}",
        ]
        if fm.created_at:
            lines.append(f"created_at: {fm.created_at}")
        if fm.updated_at:
            lines.append(f"updated_at: {fm.updated_at}")
        if fm.writing_duration_sec:
            lines.append(f"writing_duration_sec: {fm.writing_duration_sec}")
        if fm.tags:
            joined = ", ".join(self._yaml_escape(t) for t in fm.tags)
            lines.append(f"tags: [{joined}]")
        if fm.pinned:
            lines.append("pinned: true")
        lines.append("---")
        return "\n".join(lines) + "\n\n" + body.rstrip() + "\n"

    def _extract_legacy_context(self, text: str, entry_date: str) -> Optional[DayContext]:
        m = re.match(r"^---\s*\n(.*?)\n---", text, re.DOTALL)
        if not m:
            return None
        data: dict[str, str] = {}
        for line in m.group(1).splitlines():
            if ":" not in line:
                continue
            k, v = line.split(":", 1)
            data[k.strip()] = v.strip().strip('"')
        if not any(k in data for k in ("location", "weather", "temp_c", "context_source")):
            return None
        temp = None
        if data.get("temp_c"):
            try:
                temp = float(data["temp_c"])
            except ValueError:
                temp = None
        return DayContext(
            date=entry_date,
            location=data.get("location", ""),
            weather=data.get("weather", ""),
            temp_c=temp,
            context_source=data.get("context_source", "") or "desktop",
            context_updated_at=data.get("context_updated_at", "")
            or data.get("updated_at", ""),
            updated_at=data.get("context_updated_at", "") or data.get("updated_at", ""),
        )

    def _parse(self, text: str) -> tuple[str, EntryFrontMatter]:
        m = FRONT_MATTER_RE.match(text)
        if not m:
            return text, EntryFrontMatter()
        raw, body = m.group(1), m.group(2).lstrip("\n")
        data: dict[str, str] = {}
        for line in raw.splitlines():
            if ":" not in line:
                continue
            key, val = line.split(":", 1)
            data[key.strip()] = val.strip().strip('"')
        temp: Optional[float] = None
        if data.get("temp_c"):
            try:
                temp = float(data["temp_c"])
            except ValueError:
                temp = None
        duration = 0
        if data.get("writing_duration_sec"):
            try:
                duration = int(float(data["writing_duration_sec"]))
            except ValueError:
                duration = 0
        tags = self._parse_tags(data.get("tags", ""))
        fm = EntryFrontMatter(
            date=data.get("date", ""),
            title=data.get("title", ""),
            id=data.get("id", ""),
            created_at=data.get("created_at", ""),
            updated_at=data.get("updated_at", ""),
            writing_duration_sec=duration,
            tags=tags or None,
            pinned=str(data.get("pinned", "")).lower() == "true",
            location=data.get("location", ""),
            weather=data.get("weather", ""),
            temp_c=temp,
            context_source=data.get("context_source", ""),
            context_updated_at=data.get("context_updated_at", ""),
        )
        return body, fm

    @staticmethod
    def _parse_tags(raw: str) -> list[str]:
        s = (raw or "").strip()
        if not s:
            return []
        if s.startswith("[") and s.endswith("]"):
            s = s[1:-1]
        out: list[str] = []
        for part in s.split(","):
            t = part.strip().strip('"').strip("'")
            if t:
                out.append(t)
        return out

    @staticmethod
    def _yaml_escape(value: str) -> str:
        if any(c in value for c in ":#{}[],&*?|>!%@`'\"\\"):
            escaped = value.replace('"', '\\"')
            return f'"{escaped}"'
        return value

    # Legacy v1 helpers (date-keyed paths) — used only during migration detection
    def legacy_path_for(self, entry_date: str) -> Path:
        return self.diary_root / diary_md_relpath(entry_date)

    def legacy_exists(self, entry_date: str) -> bool:
        return self.legacy_path_for(entry_date).exists()

File: android/app/src/main/java/com/personaldiary/android/data/MarkdownStore.kt
package com.personaldiary.android.data

import java.io.File
import java.util.UUID

/**
 * v2 layout: diary/YYYY/MM/<id>.md
 */
class MarkdownStore(
    private val diaryRoot: File,
    private val assetsRoot: File? = null,
) {
    data class FrontMatter(
        val date: String = "",
        val title: String = "",
        val id: String = "",
        val createdAt: String = "",
        val updatedAt: String = "",
        val writingDurationSec: Int = 0,
        val tags: List<String> = emptyList(),
        val pinned: Boolean = false,
        // Legacy v1 (migration only)
        val location: String = "",
        val weather: String = "",
        val tempC: Double? = null,
        val contextSource: String = "",
        val contextUpdatedAt: String = "",
    )

    init {
        diaryRoot.mkdirs()
    }

    fun pathFor(entryId: String, entryDate: String): File {
        val parts = entryDate.split("-")
        require(parts.size == 3) { "bad date $entryDate" }
        val (y, m, _) = parts
        return File(diaryRoot, "$y/$m/$entryId.md")
    }

    fun exists(entryId: String, entryDate: String): Boolean =
        pathFor(entryId, entryDate).isFile

    fun readBody(entryId: String, entryDate: String): String {
        val f = pathFor(entryId, entryDate)
        if (!f.isFile) return ""
        val (body, _) = parse(f.readText(Charsets.UTF_8))
        return body
    }

    fun readRaw(entryId: String, entryDate: String): String {
        val f = pathFor(entryId, entryDate)
        return if (f.isFile) f.readText(Charsets.UTF_8) else ""
    }

    fun writeRaw(entryId: String, entryDate: String, markdown: String) {
        val file = pathFor(entryId, entryDate)
        file.parentFile?.mkdirs()
        file.writeText(markdown, Charsets.UTF_8)
    }

    fun readFrontMatter(entryId: String, entryDate: String): FrontMatter {
        val f = pathFor(entryId, entryDate)
        if (!f.isFile) return FrontMatter(date = entryDate, id = entryId)
        val (_, fm) = parse(f.readText(Charsets.UTF_8))
        return fm.copy(
            date = fm.date.ifBlank { entryDate },
            id = fm.id.ifBlank { entryId },
        )
    }

    fun write(
        entryId: String,
        entryDate: String,
        body: String,
        title: String = entryDate,
        createdAt: String = "",
        updatedAt: String = "",
        writingDurationSec: Int = 0,
        tags: List<String> = emptyList(),
        pinned: Boolean = false,
    ) {
        val file = pathFor(entryId, entryDate)
        file.parentFile?.mkdirs()
        val lines = mutableListOf(
            "---",
            "date: $entryDate",
            "title: ${yamlEscape(title.ifBlank { entryDate })}",
            "id: $entryId",
        )
        if (createdAt.isNotBlank()) lines += "created_at: $createdAt"
        if (updatedAt.isNotBlank()) lines += "updated_at: $updatedAt"
        if (writingDurationSec > 0) lines += "writing_duration_sec: $writingDurationSec"
        if (tags.isNotEmpty()) lines += "tags: [${tags.joinToString(", ") { yamlEscape(it) }}]"
        if (pinned) lines += "pinned: true"
        lines += "---"
        val content = lines.joinToString("\n") + "\n\n" + body.trimEnd() + "\n"
        file.writeText(content, Charsets.UTF_8)
    }

    fun listNoteIds(): List<Pair<String, String>> {
        if (!diaryRoot.exists()) return emptyList()
        val found = mutableListOf<Pair<String, String>>()
        diaryRoot.walkTopDown()
            .filter { it.isFile && it.extension == "md" }
            .forEach { path ->
                if (DATE_FILE_RE.matches(path.name)) return@forEach
                val stem = path.nameWithoutExtension
                if (!UUID_RE.matches(stem)) return@forEach
                val text = path.readText(Charsets.UTF_8)
                val (_, fm) = parse(text)
                val entryDate = fm.date
                if (entryDate.isBlank()) return@forEach
                found += stem to entryDate
            }
        return found.sortedByDescending { it.second }
    }

    fun migrateV1Layout(dayStore: DayStore): Int {
        if (!diaryRoot.exists()) return 0
        var count = 0
        val legacyFiles = diaryRoot.walkTopDown()
            .filter { it.isFile && DATE_FILE_RE.matches(it.name) }
            .sortedBy { it.path }
            .toList()
        for (path in legacyFiles) {
            val entryDate = path.nameWithoutExtension
            val text = path.readText(Charsets.UTF_8)
            val (body, fm) = parse(text)
            val entryId = fm.id.ifBlank { UUID.randomUUID().toString() }
            val date = fm.date.ifBlank { entryDate }

            val dayCtx = extractLegacyContext(text, entryDate)
            if (dayCtx != null && dayCtx.hasContext) {
                dayStore.mergeWrite(dayCtx)
            }

            val newBody = body.replace("assets/$entryDate/", "assets/$entryId/")
            writeRaw(entryId, date, renderV2(newBody, fm.copy(id = entryId, date = date)))

            if (assetsRoot != null) {
                val oldAssets = File(assetsRoot, entryDate)
                if (oldAssets.isDirectory) {
                    val newAssets = File(assetsRoot, entryId).also { it.mkdirs() }
                    oldAssets.listFiles()?.filter { it.isFile }?.forEach { ap ->
                        val target = File(newAssets, ap.name)
                        if (!target.exists()) ap.renameTo(target)
                    }
                    oldAssets.listFiles()?.takeIf { it.isEmpty() }?.let { oldAssets.delete() }
                }
            }

            path.delete()
            count++
        }
        return count
    }

    fun parse(text: String): Pair<String, FrontMatter> {
        if (!text.startsWith("---")) return text to FrontMatter()
        val end = text.indexOf("\n---", 3)
        if (end < 0) return text to FrontMatter()
        val raw = text.substring(3, end).trim('\n')
        val body = text.substring(end + 4).trimStart('\n')
        val map = mutableMapOf<String, String>()
        raw.lineSequence().forEach { line ->
            val idx = line.indexOf(':')
            if (idx > 0) {
                map[line.substring(0, idx).trim()] = line.substring(idx + 1).trim().trim('"')
            }
        }
        val temp = map["temp_c"]?.toDoubleOrNull()
        val duration = map["writing_duration_sec"]?.toDoubleOrNull()?.toInt() ?: 0
        return body to FrontMatter(
            date = map["date"].orEmpty(),
            title = map["title"].orEmpty(),
            id = map["id"].orEmpty(),
            createdAt = map["created_at"].orEmpty(),
            updatedAt = map["updated_at"].orEmpty(),
            writingDurationSec = duration,
            tags = parseTags(map["tags"].orEmpty()),
            pinned = map["pinned"].orEmpty().equals("true", ignoreCase = true),
            location = map["location"].orEmpty(),
            weather = map["weather"].orEmpty(),
            tempC = temp,
            contextSource = map["context_source"].orEmpty(),
            contextUpdatedAt = map["context_updated_at"].orEmpty(),
        )
    }

    private fun renderV2(body: String, fm: FrontMatter): String {
        val title = fm.title.ifBlank { fm.date.ifBlank { "untitled" } }
        val lines = mutableListOf(
            "---",
            "date: ${fm.date}",
            "title: ${yamlEscape(title)}",
            "id: ${fm.id}",
        )
        if (fm.createdAt.isNotBlank()) lines += "created_at: ${fm.createdAt}"
        if (fm.updatedAt.isNotBlank()) lines += "updated_at: ${fm.updatedAt}"
        if (fm.writingDurationSec > 0) lines += "writing_duration_sec: ${fm.writingDurationSec}"
        if (fm.tags.isNotEmpty()) {
            lines += "tags: [${fm.tags.joinToString(", ") { yamlEscape(it) }}]"
        }
        if (fm.pinned) lines += "pinned: true"
        lines += "---"
        return lines.joinToString("\n") + "\n\n" + body.trimEnd() + "\n"
    }

    private fun parseTags(raw: String): List<String> {
        val s = raw.trim()
        if (s.isEmpty()) return emptyList()
        val inner = when {
            s.startsWith("[") && s.endsWith("]") -> s.substring(1, s.length - 1)
            else -> s
        }
        return inner.split(',')
            .map { it.trim().trim('"').trim('\'') }
            .filter { it.isNotEmpty() }
            .distinct()
    }

    private fun extractLegacyContext(text: String, entryDate: String): DayContext? {
        if (!text.startsWith("---")) return null
        val end = text.indexOf("\n---", 3)
        if (end < 0) return null
        val raw = text.substring(3, end)
        val map = mutableMapOf<String, String>()
        raw.lineSequence().forEach { line ->
            val idx = line.indexOf(':')
            if (idx > 0) map[line.substring(0, idx).trim()] = line.substring(idx + 1).trim().trim('"')
        }
        if (!map.keys.any { it in setOf("location", "weather", "temp_c", "context_source") }) {
            return null
        }
        return DayContext(
            date = entryDate,
            location = map["location"].orEmpty(),
            weather = map["weather"].orEmpty(),
            tempC = map["temp_c"]?.toDoubleOrNull(),
            contextSource = map["context_source"].orEmpty().ifBlank { "desktop" },
            contextUpdatedAt = map["context_updated_at"].orEmpty().ifBlank {
                map["updated_at"].orEmpty()
            },
            updatedAt = map["context_updated_at"].orEmpty().ifBlank {
                map["updated_at"].orEmpty()
            },
        )
    }

    private fun yamlEscape(value: String): String {
        return if (value.any { it in ":#{}[],&*?|>!%@`'\"" || it == '\\' }) {
            "\"${value.replace("\"", "\\\"")}\""
        } else value
    }

    companion object {
        private val DATE_FILE_RE = Regex("""^\d{4}-\d{2}-\d{2}\.md$""")
        private val UUID_RE = Regex(
            """^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"""
        )
    }
}
